src/bin/client.rs

```diff
@@ -0,0 +1,227 @@
+        verifier.verify(sig)
+    }
+}
+
+fn sha_512(prefix: &[u8], data: &[u8]) -> Vec<u8> {
+    let mut ctx = digest::Context::new(&digest::SHA512);
+    ctx.update(prefix);
+    ctx.update(data);
+    Vec::from(ctx.finish().as_ref())
+}
+
+fn main() {
+    let matches = App::new("roughenough client")
+                      .version("0.1.1")
+                      .arg(Arg::with_name("host")
+                           .required(true)
+                           .help("The Roughtime server to connect to")
+                           .takes_value(true))
+                      .arg(Arg::with_name("port")
+                           .required(true)
+                           .help("The Roughtime server port to connect to")
+                           .takes_value(true))
+                      .arg(Arg::with_name("public-key")
+                           .short("p")
+                           .long("public-key")
+                           .takes_value(true)
+                           .help("The server public key used to validate responses. If unset, no validation will be performed"))
+                      .arg(Arg::with_name("time-format")
+                           .short("f")
+                           .long("time-format")
+                           .takes_value(true)
+                           .help("The strftime format string used to print the time recieved from the server")
+                           .default_value("%b %d %Y %H:%M:%S")
+                        )
+                      .arg(Arg::with_name("num-requests")
+                           .short("n")
+                           .long("num-requests")
+                           .takes_value(true)
+                           .help("The number of requests to make to the server (each from a different source port). This is mainly useful for testing batch response handling")
+                           .default_value("1")
+                        )
+                      .get_matches();
+
+    let host = matches.value_of("host").unwrap();
+    let port = value_t_or_exit!(matches.value_of("port"), u16);
+    let num_requests = value_t_or_exit!(matches.value_of("num-requests"), u16) as usize;
+    let pub_key = matches.value_of("public-key").map(|pkey| hex::decode(pkey).expect("Error parsing public key!"));
+    let time_format = matches.value_of("time-format").unwrap();
+
+    println!("Requesting time from: {:?}:{:?}", host, port);
+
+    let addrs: Vec<_> = (host, port).to_socket_addrs().unwrap().collect();
+
+    let mut requests = Vec::with_capacity(num_requests);
+
+    for _ in 0..num_requests {
+        let nonce = create_nonce();
+        let mut socket = UdpSocket::bind("0.0.0.0:0").expect("Couldn't open UDP socket");
+        let request = make_request(&nonce);
+
+        requests.push((nonce, request, socket));
+    }
+
+    for &mut (_, ref request, ref mut socket) in requests.iter_mut() {
+        socket.send_to(request, addrs.as_slice()).unwrap();
```

Comment:
How about using `parse()` for getting the single `SockAddr` for the `send_to`?

```rust
let addr: SockAddr = format!("{}:{}", host, port).parse().unwrap("Couldn't create destination socket");
```

When I saw the wrapping`Vec` I initially thought there was a bug given `send_to` uses only the first address. But then I realized there is only the single dest address, so it was correct, just not obvious.